### PR TITLE
feat: add guardian proto

### PR DESCRIPTION
**Pull request by rahmatrhd**

### CODE REVIEW

**Review by ravisuhag — CHANGES_REQUESTED** [review]

File: odpf/guardian/guardian.proto [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/odpf/proton/guardian/go_package";
+option java_multiple_files = true;
+option java_package = "io.odpf.proton.guardian";
+option java_outer_classname = "ServiceManager";
+
+// These annotations are used when generating the OpenAPI file.
+option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_swagger) = {
```
[the remark this answers is not in the document]
Comment (rahmatrhd) [reply odpf/guardian/guardian.proto:15]: https://github.com/grpc-ecosystem/grpc-gateway/pull/2147 the support for v3 still under development

File: odpf/guardian/guardian.proto [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+        };
+    }
+
+    rpc ListApprovals(ListApprovalsRequest) returns (ListApprovalsResponse) {
+        option (google.api.http) = {
+            get: "/appeals/approvals"
+        };
+    }
+
+    rpc MakeAction(MakeActionRequest) returns (MakeActionResponse) {
```
[the remark this answers is not in the document]
Comment (rahmatrhd) [reply odpf/guardian/guardian.proto]: renamed to `UpdateApproval`

**Comment by ravisuhag** [comment]
LGTM. @kushsharma Can you also cross-check for any conventions we are following in proton?

File: odpf/guardian/guardian.proto [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+    Action action = 3;
+}
+
+message UpdateApprovalResponse {
+    Appeal appeal = 1;
+}
+
+// Provider contains information about external data provider such as BigQuery, Metabase, etc., credentials, policy, and allowed roles 
+message Provider {
+    uint32 id = 1;
```
Comment (kushsharma) [inline odpf/guardian/guardian.proto:253]: I see we are using int ids everywhere, are these DB auto-increment ids? Why not use UUID?

File: odpf/guardian/guardian.proto [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    string type = 2;
+    string urn = 3;
+
+    message ProviderConfig {
+        string type = 1;
+        string urn = 2;
+        google.protobuf.Struct labels = 3;
```
Comment (kushsharma) [inline odpf/guardian/guardian.proto:260]: I see labels as `key,value` pairs of string, any reason why we chose Struct for this? We should limit the usage of `Struct` to only places where we know for sure we can't use concrete types I think as it adds inconvenience while parsing it.

**Review by kushsharma — CHANGES_REQUESTED** [review]

**Comment by kushsharma** [comment]
@ravisuhag are we not versioning our REST APIs? Not versioning protobufs is fine I think as they get imported and get versioned while importing but REST APIs have no convention of knowing if something has breaking changes right?

File: odpf/guardian/guardian.proto (hunk 3 above)
Comment (rahmatrhd) [reply odpf/guardian/guardian.proto:253]: Yes DB auto-increment, no specific reason why we're not using UUID, just following traditional ways for creating entity. What would make it better to use UUID?

File: odpf/guardian/guardian.proto (hunk 4 above)
Comment (rahmatrhd) [reply odpf/guardian/guardian.proto:260]: The type we have for labels in guardian right now is `map[string]interface{}`, using Struct to fulfill that.

Comment (ravisuhag) [reply odpf/guardian/guardian.proto:260]: I think this is better moved to [string][string] labels should not be In any other type.

**Comment by ravisuhag** [comment]
+1 to versioning. Shall we right away start with v1 as folder as well and move protos to that? or any other better idea?

File: odpf/guardian/guardian.proto (hunk 4 above)
Comment (rahmatrhd) [reply odpf/guardian/guardian.proto:260]: changing the proto type for labels to `map<string, string>`. Will change the labels type in guardian to `map[string]string` as well

**Comment by ravisuhag** [comment]
let's skip `/api ` if needed service should handle it.

File: odpf/guardian/guardian.proto [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+        version: "0.1";
+    };
+    external_docs: {
+        description: "Guardian server";
+    }
+    schemes: HTTP;
+};
+
+// WARNING: This is still in active development and can have breaking changes
+service Guardian {
```
Comment (kushsharma) [inline odpf/guardian/guardian.proto:26]: It's a general guideline to add `Service` suffix to services to distinguish them from proto messages.

File: odpf/guardian/guardian.proto (hunk 3 above)
Comment (kushsharma) [reply odpf/guardian/guardian.proto:253]: - can't enumerate values
- can't infer how many values are present in each table
- don't have to talk to the database to construct a value, useful in certain ORM situations
- perfectly indexable by most dbs
- easier to shard when we need to distribute data into multiple dbs
- unique across multiple tables/databases/space-time continuum
- costly on size but storage is cheap

**Comment by kushsharma** [comment]
Yeah, we can version protobufs at directory/package level as well although once we have stable services, function definition hardly changes, only message fields do and that can be handled by only introducing new fields with new indices and supporting old fields as well. But yeah I am onboard with directory level versioning as well.
I need to refactor `optimus` protobufs though, went with the simplest one when started and now it has so many things going on.
Like this
```
.
└── proto
    └── foo
        └── bar
            ├── bat
            │   └── v1
            │       └── bat.proto // package foo.bar.bat.v1
            └── baz
                └── v1
                    ├── baz.proto         // package foo.bar.baz.v1
                    └── baz_service.proto // package foo.bar.baz.v1
```

**Comment by ravisuhag** [comment]
@rahmatrhd lets make the discussed changes and then we can merge.

**Comment by ravisuhag** [comment]
@rahmatrhd Are we moving protos to V1 folder?

**Comment by kushsharma** [comment]
> @rahmatrhd @kushsharma Are we moving protos to V1 folder?

yeah we should

**Comment by rohilsurana** [comment]
`package odpf.guardian;` => `package odpf.guardian.v1;`
`option go_package = "github.com/odpf/proton/guardian";` => `option go_package = "github.com/odpf/proton/guardian/v1;guardianv1";`

same goes for the directory as well.

**Comment by rohilsurana** [comment]
Let's follow semver for `openapiv2_swagger.info.version`.

**Pull request merged**